File: 04-Data-Structures/code/python/seb_ds/radix_tree.py
"""Compressed trie (radix tree / PATRICIA-style). Mirrors `typescript/src/radixTree.ts`.

Deletion is intentionally omitted (optional per lab scope).
"""

from __future__ import annotations

from typing import Dict, NamedTuple, Optional


class _RadixNode:
    __slots__ = ("children", "is_end")

    def __init__(self) -> None:
        self.children: Dict[str, "_Edge"] = {}
        self.is_end = False


class _Edge(NamedTuple):
    label: str
    node: _RadixNode


class RadixTree:
    def __init__(self) -> None:
        self._root = _RadixNode()

    def insert(self, word: str) -> None:
        self._insert_rec(self._root, word)

    def _insert_rec(self, node: _RadixNode, word: str) -> None:
        if len(word) == 0:
            node.is_end = True
            return
        first_char = word[0]
        edge = node.children.get(first_char)
        if edge is None:
            leaf = _RadixNode()
            leaf.is_end = True
            node.children[first_char] = _Edge(word, leaf)
            return
        cp = self._common_prefix_len(edge.label, word)
        if cp == len(edge.label):
            self._insert_rec(edge.node, word[cp:])
            return
        split_node = _RadixNode()
        old_label_rest = edge.label[cp:]
        split_node.children[old_label_rest[0]] = _Edge(old_label_rest, edge.node)
        new_label = edge.label[:cp]
        node.children[first_char] = _Edge(new_label, split_node)
        word_rest = word[cp:]
        if len(word_rest) == 0:
            split_node.is_end = True
        else:
            self._insert_rec(split_node, word_rest)

    def _common_prefix_len(self, a: str, b: str) -> int:
        i = 0
        while i < len(a) and i < len(b) and a[i] == b[i]:
            i += 1
        return i

    def contains(self, word: str) -> bool:
        located = self._locate_exact(word)
        return located is not None and located.is_end

    def starts_with(self, prefix: str) -> bool:
        if len(prefix) == 0:
            return True
        return self._locate_prefix(self._root, prefix)

    def _locate_prefix(self, node: _RadixNode, prefix: str) -> bool:
        if len(prefix) == 0:
            return True
        edge = node.children.get(prefix[0])
        if edge is None:
            return False
        cp = self._common_prefix_len(edge.label, prefix)
        if cp == len(prefix):
            return True
        if cp == len(edge.label):
            return self._locate_prefix(edge.node, prefix[cp:])
        return False

    def _locate_exact(self, word: str) -> Optional[_RadixNode]:
        node = self._root
        remaining = word
        while len(remaining) > 0:
            edge = node.children.get(remaining[0])
            if edge is None:
                return None
            cp = self._common_prefix_len(edge.label, remaining)
            if cp < len(edge.label):
                return None
            remaining = remaining[cp:]
            node = edge.node
        return node

    def check_invariants(self) -> None:
        pass

File: 04-Data-Structures/code/typescript/src/radixTree.ts
class RadixNode {
  children: Map<string, { label: string; node: RadixNode }> = new Map();
  isEnd = false;
}

/**
 * Compressed trie (radix tree / PATRICIA-style): edges carry whole label
 * strings instead of one node per character, collapsing long unbranching
 * chains. Deletion is intentionally omitted (optional per lab scope).
 */
export class RadixTree {
  private root = new RadixNode();

  insert(word: string): void {
    this.insertRec(this.root, word);
  }

  private insertRec(node: RadixNode, word: string): void {
    if (word.length === 0) {
      node.isEnd = true;
      return;
    }
    const firstChar = word[0];
    const edge = node.children.get(firstChar);
    if (!edge) {
      const leaf = new RadixNode();
      leaf.isEnd = true;
      node.children.set(firstChar, { label: word, node: leaf });
      return;
    }
    const cp = this.commonPrefixLen(edge.label, word);
    if (cp === edge.label.length) {
      this.insertRec(edge.node, word.slice(cp));
      return;
    }
    const splitNode = new RadixNode();
    const oldLabelRest = edge.label.slice(cp);
    splitNode.children.set(oldLabelRest[0], { label: oldLabelRest, node: edge.node });
    const newLabel = edge.label.slice(0, cp);
    node.children.set(firstChar, { label: newLabel, node: splitNode });
    const wordRest = word.slice(cp);
    if (wordRest.length === 0) splitNode.isEnd = true;
    else this.insertRec(splitNode, wordRest);
  }

  private commonPrefixLen(a: string, b: string): number {
    let i = 0;
    while (i < a.length && i < b.length && a[i] === b[i]) i++;
    return i;
  }

  contains(word: string): boolean {
    const located = this.locateExact(word);
    return located !== null && located.node.isEnd;
  }

  startsWith(prefix: string): boolean {
    if (prefix.length === 0) return true;
    return this.locatePrefix(this.root, prefix);
  }

  private locatePrefix(node: RadixNode, prefix: string): boolean {
    if (prefix.length === 0) return true;
    const edge = node.children.get(prefix[0]);
    if (!edge) return false;
    const cp = this.commonPrefixLen(edge.label, prefix);
    if (cp === prefix.length) return true;
    if (cp === edge.label.length) return this.locatePrefix(edge.node, prefix.slice(cp));
    return false;
  }

  private locateExact(word: string): { node: RadixNode } | null {
    let node = this.root;
    let remaining = word;
    while (remaining.length > 0) {
      const edge = node.children.get(remaining[0]);
      if (!edge) return null;
      const cp = this.commonPrefixLen(edge.label, remaining);
      if (cp < edge.label.length) return null;
      remaining = remaining.slice(cp);
      node = edge.node;
    }
    return { node };
  }

  checkInvariants(): void {
    // Structural by construction; see Trie.checkInvariants for rationale.
  }
}
